pkg/maps/l2v6respondermap/l2_responder_map6.go

```diff
@@ -0,0 +1,135 @@
+	Create(ip netip.Addr, ifIndex uint32) error
+	Lookup(ip netip.Addr, ifIndex uint32) (*l2respondermap.L2ResponderStats, error)
+	Delete(ip netip.Addr, ifIndex uint32) error
+	IterateWithCallback(cb IterateCallback) error
+}
+
+func NewMap(lifecycle cell.Lifecycle, logger *slog.Logger) Map {
+	return newMap(lifecycle, DefaultMaxEntries, logger)
+}
+
+type l2V6ResponderMap struct {
+	*ebpf.Map
+}
+
+func newMap(lifecycle cell.Lifecycle, maxEntries int, logger *slog.Logger) *l2V6ResponderMap {
+	outerMap := &l2V6ResponderMap{}
+
+	lifecycle.Append(cell.Hook{
+		OnStart: func(hc cell.HookContext) error {
+			var (
+				m   *ebpf.Map
+				err error
+			)
+
+			if m, err = ebpf.LoadRegisterMap(logger, MapName); err != nil {
+				m = ebpf.NewMap(logger, &ebpf.MapSpec{
+					Name:       MapName,
+					Type:       ebpf.Hash,
+					KeySize:    uint32(unsafe.Sizeof(L2V6ResponderKey{})),
+					ValueSize:  uint32(unsafe.Sizeof(l2respondermap.L2ResponderStats{})),
+					MaxEntries: uint32(maxEntries),
+					Flags:      unix.BPF_F_NO_PREALLOC,
+					Pinning:    ebpf.PinByName,
+				})
+				if err := m.OpenOrCreate(); err != nil {
+					return err
+				}
+			}
+
+			outerMap.Map = m
+
+			return nil
+		},
+	})
+
+	return outerMap
+}
+
+// Create creates a new entry for the given IP and IfIndex tuple.
+func (m *l2V6ResponderMap) Create(ip netip.Addr, ifIndex uint32) error {
+	key := newL2V6ResponderKey(ip, ifIndex)
+	return m.Map.Put(key, l2respondermap.L2ResponderStats{})
+}
+
+// Delete deletes the entry associated with the provided IP and IfIndex tuple.
+func (m *l2V6ResponderMap) Delete(ip netip.Addr, ifIndex uint32) error {
+	key := newL2V6ResponderKey(ip, ifIndex)
+	return m.Map.Delete(key)
+}
+
+// Lookup returns the stats object associated with the provided IP and IfIndex tuple.
+func (m *l2V6ResponderMap) Lookup(ip netip.Addr, ifIndex uint32) (*l2respondermap.L2ResponderStats, error) {
+	key := newL2V6ResponderKey(ip, ifIndex)
+	val := l2respondermap.L2ResponderStats{}
+
+	err := m.Map.Lookup(&key, &val)
+
+	return &val, err
+}
+
+// IterateCallback represents the signature of the callback function
+// expected by the IterateWithCallback method, which in turn is used to iterate
+// all the keys/values of a L2 responder map.
+type IterateCallback func(*L2V6ResponderKey, *l2respondermap.L2ResponderStats)
+
+// IterateWithCallback iterates through all the keys/values of a L2 responder map,
+// passing each key/value pair to the cb callback.
+func (m *l2V6ResponderMap) IterateWithCallback(cb IterateCallback) error {
+	return m.Map.IterateWithCallback(&L2V6ResponderKey{}, &l2respondermap.L2ResponderStats{},
+		func(k, v interface{}) {
+			key := k.(*L2V6ResponderKey)
+			value := v.(*l2respondermap.L2ResponderStats)
+			cb(key, value)
+		},
+	)
+}
+
+// L2V6ResponderKey implements the bpf.MapKey interface.
+//
+// Must be in sync with struct l2_responder_v4_key in <bpf/lib/maps.h>
+type L2V6ResponderKey struct {
+	IP      types.IPv6 `align:"address"`
```

Comment:
The corresponding definition in `bpf/lib/l2_responder.h` defined this field with the name `ip6`, which tells me that the align checker is not validating that the Go and C structures have equivalent alignment (or maybe something has changed there since I last touched alignchecker?).
```suggestion
	IP      types.IPv6 `align:"ip6"`
```

Reply:
@cilium/sig-datapath we might want to consider looking into `bpf/bpf_alignchecker.c` to see whether we're still ensuring alignment between Go and C versions of these structures.

Reply:
FWIW, in `main` v4 is also not matching (`ip4` vs `ip`):

https://github.com/cilium/cilium/blob/main/bpf/lib/l2_responder.h#L11
https://github.com/cilium/cilium/blob/main/pkg/maps/l2respondermap/l2_responder_map4.go#L119

I will fix it as part of adding the map commit, but I think the alignchecker discussion should be stubbed somewhere else.

Reply:
Yep makes sense. I'll mark this thread as "resolved" as it doesn't need to be addressed in this PR. I just think that the codeowner group should consider how we're ensuring proper alignment of these fields going forward; it seems like somewhere along the way we stopped enforcing the same mechanism for alignment. I know that early on in the project, it was semi-common for alignment to differ and this wasted a lot of contributor time to figure out, which was why we put the automated check in place. I guess that context was not passed on to newer reviewers as they joined the group.